devito/passes/clusters/blocking.py

```diff
@@ -157,3 +160,102 @@ def decompose(ispace, d, block_dims):
     directions.update({bd: ispace.directions[d] for bd in block_dims})
 
     return IterationSpace(intervals, sub_iterators, directions)
+
+
+class Skewing(Queue):
+
+    def __init__(self, options):
+
+        super(Skewing, self).__init__()
+
+    @timed_pass(name='skewing')
+    def process(self, clusters):
+        return super(Skewing, self).process(clusters)
+
+    def _process_fdta(self, clusters, level, prefix=None):
+
+        return super(Skewing, self)._process_fdta(clusters, level, prefix)
+
+    def callback(self, clusters, prefix):
+        if not prefix:
+            return clusters
+
+        d = prefix[-1].dim
+
+        processed = []
+        for c in clusters:
+            if PARALLEL in c.properties[d] and not d.symbolic_incr.is_Symbol:
+                ispace, exprs = skew(c.ispace, c.exprs, c.properties, d)
+                processed.append(c.rebuild(exprs=exprs, ispace=ispace,
+                                           properties=c.properties))
+            else:
+                processed.append(c)
+
+        return processed
+
+
+def skew(ispace, exprs, properties, d):
+
+    """
+    Create a new IterationSpace and skew expressions
+    Skew the accesses along a SEQUENTIAL Dimension.
+    Reference:
+    https://link.springer.com/article/10.1007%2FBF01407876
+    Example:
+
+    Transform
+
+    for i = 2, n-1
+        for j = 2, m-1
+            a[i,j] = (a[a-1,j] + a[i,j-1] + a[i+1,j] + a[i,j+1]) / 4
+        end for
+    end for
+
+    to
+
+    for i = 2, n-1
+        for j = 2+i, m-1+i
+            a[i,j-i] = (a[a-1,j-i] + a[i,j-1-i] + a[i+1,j-i] + a[i,j+1-i]) / 4
+        end for
+    end for
+    """
+
+    # What dimensions should we skew against?
+    # e.g. SEQUENTIAL Dimensions (Usually time in FD solvers, but not only limited
+    # to this)
+    # What dimensions should be skewed?
+    # e.g. PARALLEL dimensions (x, y, z) but not blocking ones (x_blk0, y_blk0)
+    # Iterate over the iteration space and assign dimensions to skew or skewable
+    # depending on their properties
+
+    # Skew dim will not be none here:
+    # Initializing a default skewed dim index position in loop
+
+    mapper, intervals, processed = {}, [], []
+
+    # Identify a skew_dim and its position
+    skew_flag = False
```

Comment:
this piece of code is flawed in many ways. First of all, it will fail if no SEQUENTIAL Dimensions are detected, since skew_dim is only defined inside the loop body.

Then, you're using three variables to carry one piece of information...  that's such a useless complication

I _think_ you could also get away with a _single_ `for i in ispace` loop -- you only need to track the first SEQUENTIAL Dimension, then you skew the inner PARALLEL loops afterwards

Reply:
I think it is better now, according to your comments